File: runtime/forge/internal/dsh/patch.go
package dsh

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

// ProtocolVersion is the @deepseek-ai/dsh client protocol this package
// renders patches for.
const ProtocolVersion = "0.1.0-rc.6"

// Loader overlay row ids pinned by the rc.6 protocol. The providers and the
// selected default model are targeted by id; the bridge plugin and each MCP
// server are injected rows.
const (
	// llmPIAIProviderRowID targets the injected provider group by id.
	llmPIAIProviderRowID = "llm-pi-ai"
	// agentDefaultModelRowID targets the selected default model by id.
	agentDefaultModelRowID = "agent-default-model"
	// BridgeRowID is the loader overlay insert id for the Forge bridge plugin.
	BridgeRowID = "forge-dsh-bridge"
	// MCPClientRowID is the loader overlay insert id for injected MCP clients.
	MCPClientRowID = "@deepseek-ai/dsh-mcp-client"
)

// ToolCapability is a capability projection requested at launch time. Only
// MCP projection is supported in this release; any other capability fails
// loudly.
type ToolCapability string

const (
	// ToolMCP is the supported MCP capability projection.
	ToolMCP ToolCapability = "mcp"
	// ToolBash is unsupported; requesting it fails loudly.
	ToolBash ToolCapability = "bash"
	// ToolFS is an unsupported non-MCP tool; requesting it fails loudly.
	ToolFS ToolCapability = "fs"
)

// MCPTransport is the transport of an MCP capability row.
type MCPTransport string

const (
	// MCPTransportStdio renders a stdio MCP server row.
	MCPTransportStdio MCPTransport = "stdio"
	// MCPTransportStreamableHTTP renders a streamable-http MCP server row.
	MCPTransportStreamableHTTP MCPTransport = "streamable-http"
)

// MCPServer is a single projected MCP capability row. Env overrides never
// carry credentials from this package.
type MCPServer struct {
	Name      string
	Transport MCPTransport
	Command   string
	Args      []string
	Env       []string
}

// PatchInput is the complete secret-free input to RenderPatch.
type PatchInput struct {
	Providers        []Provider
	SelectedModel    string
	MCPServers       []MCPServer
	Tools            []ToolCapability
	Version          string
	BridgePluginPath string // absolute plugin path; empty emits no bridge insert row
}

var yamlPlainRe = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._@/\-]*$`)

func yamlStr(s string) string {
	if yamlPlainRe.MatchString(s) {
		return s
	}
	return fmt.Sprintf("%q", s)
}

// RenderPatch renders the deterministic, secret-free DSH loader overlay for
// the @deepseek-ai/dsh 0.1.0-rc.6 protocol. The overlay is a top-level YAML
// array: an llm-pi-ai row whose config.providers dict is keyed by canonical
// route (id without the llm-pi-ai. prefix), an optional agent-default-model
// row, and bridge/MCP insert rows. Every credential is a literal unquoted
// !!js process.env.<ENV> tag, never a value, so the file never holds a
// credential. The native deepseek-official routes are never re-emitted.
// agent-default-model is only ever added for an explicit SelectedModel. MCP
// capability rows are rendered for stdio and streamable-http transports;
// unsupported tools or Bash additions are rejected loudly.
func RenderPatch(in PatchInput) ([]byte, error) {
	if in.Version == "" {
		in.Version = ProtocolVersion
	}
	providers := cloneProviders(in.Providers)

	for _, tool := range in.Tools {
		if tool != ToolMCP {
			return nil, fmt.Errorf("dsh: unsupported tool capability %q: only mcp projection is supported", tool)
		}
	}

	selected, err := resolveSelection(providers, in.SelectedModel)
	if err != nil {
		return nil, err
	}

	var b strings.Builder
	b.WriteString("# forge dsh patch (generated; secret-free)\n")

	// llm-pi-ai loader row: providers dict keyed by canonical route.
	b.WriteString("- id: " + yamlStr(llmPIAIProviderRowID) + "\n")
	b.WriteString("  config:\n")
	b.WriteString("    providers:\n")
	for _, p := range providers {
		routeKey := RouteKey(p)
		b.WriteString("      " + yamlStr(routeKey) + ":\n")
		b.WriteString("        displayName: " + yamlStr(displayNameFor(routeKey)) + "\n")
		b.WriteString("        api: " + yamlStr(string(p.APIType)) + "\n")
		b.WriteString("        apiKeyEnv: " + yamlStr(p.APIKeyEnv) + "\n")
		b.WriteString("        baseURL: " + yamlStr(NormalizeBaseURL(p.BaseURL)) + "\n")
		b.WriteString("        models:\n")
		for _, m := range p.Models {
			b.WriteString("          - id: " + yamlStr(m.ID) + "\n")
			if strings.TrimSpace(m.Label) != "" {
				b.WriteString("            name: " + yamlStr(m.Label) + "\n")
			}
			if m.ContextWindow > 0 {
				b.WriteString("            contextWindow: " + fmt.Sprintf("%d", m.ContextWindow) + "\n")
			}
			if m.MaxTokens > 0 {
				b.WriteString("            maxTokens: " + fmt.Sprintf("%d", m.MaxTokens) + "\n")
			}
			if m.Reasoning {
				b.WriteString("            reasoningEfforts:\n")
				b.WriteString("              off:\n")
				b.WriteString("              high: high\n")
			}
		}
		if headerNames := sortedNonAuthHeaders(p.Headers); len(headerNames) > 0 {
			b.WriteString("        headers:\n")
			for _, name := range headerNames {
				b.WriteString("          " + yamlStr(name) + ": " + envRefTag(p.Headers[name]) + "\n")
			}
		}
	}

	if selected != "" {
		pid, mid, _ := splitSelection(selected)
		b.WriteString("- id: " + yamlStr(agentDefaultModelRowID) + "\n")
		b.WriteString("  config:\n")
		b.WriteString("    provider: " + yamlStr(pid) + "\n")
		b.WriteString("    model: " + yamlStr(mid) + "\n")
	}

	rows, err := RenderInsertRows(in.BridgePluginPath, in.MCPServers)
	if err != nil {
		return nil, err
	}
	b.Write(rows)
	return []byte(b.String()), nil
}

// RenderInsertRows renders the loader overlay insert rows for the bridge
// plugin and MCP servers as a YAML fragment appended to a base overlay. The
// bridge row is emitted only when bridgePluginPath is non-empty; MCP rows are
// sorted by server name and carry env/header refs as unquoted !!js tags.
func RenderInsertRows(bridgePluginPath string, servers []MCPServer) ([]byte, error) {
	for _, srv := range servers {
		if srv.Transport != MCPTransportStdio && srv.Transport != MCPTransportStreamableHTTP {
			return nil, fmt.Errorf("dsh: unsupported mcp transport %q", srv.Transport)
		}
	}
	sorted := append([]MCPServer(nil), servers...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Name < sorted[j].Name })

	var b strings.Builder
	if strings.TrimSpace(bridgePluginPath) != "" {
		b.WriteString("- insert:\n")
		b.WriteString("    id: " + yamlStr(BridgeRowID) + "\n")
		b.WriteString("    name: " + yamlStr(bridgePluginPath) + "\n")
	}
	for _, srv := range sorted {
		b.WriteString("- insert:\n")
		b.WriteString("    id: " + yamlStr(MCPClientRowID) + "\n")
		b.WriteString("    serverName: " + yamlStr(srv.Name) + "\n")
		b.WriteString("    transport: " + yamlStr(string(srv.Transport)) + "\n")
		b.WriteString("    config:\n")
		if srv.Transport == MCPTransportStdio {
			b.WriteString("      command: " + yamlStr(srv.Command) + "\n")
			if len(srv.Args) > 0 {
				b.WriteString("      args:\n")
				for _, a := range srv.Args {
					b.WriteString("        - " + yamlStr(a) + "\n")
				}
			}
		}
		if len(srv.Env) > 0 {
			env := append([]string(nil), srv.Env...)
			sort.Strings(env)
			b.WriteString("      env:\n")
			for _, e := range env {
				key, value, ok := strings.Cut(e, "=")
				if !ok {
					continue
				}
				b.WriteString("        " + yamlStr(key) + ": " + yamlValue(value) + "\n")
			}
		}
	}
	return []byte(b.String()), nil
}

// envRefTag is an unquoted YAML tag dereferencing an env var at load time.
func envRefTag(envName string) string {
	return "!!js process.env." + envName
}

// yamlValue renders a config value, passing through explicit !!js tags raw so
// they stay unquoted.
func yamlValue(v string) string {
	if strings.HasPrefix(v, "!!js ") {
		return v
	}
	return yamlStr(v)
}

// displayNameFor derives a stable human-facing route display name.
func displayNameFor(routeKey string) string {
	words := strings.Split(routeKey, "-")
	for i, w := range words {
		if w == "" {
			continue
		}
		words[i] = strings.ToUpper(w[:1]) + w[1:]
	}
	return strings.Join(words, " ")
}

// sortedNonAuthHeaders returns the sorted header names of p excluding
// Authorization, which is handled solely by apiKeyEnv.
func sortedNonAuthHeaders(headers map[string]string) []string {
	names := make([]string, 0, len(headers))
	for name := range headers {
		if strings.EqualFold(name, "Authorization") {
			continue
		}
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

func cloneProviders(src []Provider) []Provider {
	if src == nil {
		src = InjectedProviders
	}
	out := make([]Provider, 0, len(src))
	for _, p := range src {
		cp := p
		cp.Models = append([]Model(nil), p.Models...)
		if p.Headers != nil {
			headers := make(map[string]string, len(p.Headers))
			for k, v := range p.Headers {
				headers[k] = v
			}
			cp.Headers = headers
		}
		out = append(out, cp)
	}
	return out
}

func resolveSelection(providers []Provider, sel string) (string, error) {
	if sel == "" {
		return "", nil
	}
	pid, mid, err := splitSelection(sel)
	if err != nil {
		return "", err
	}
	for _, p := range providers {
		if p.ID != pid {
			continue
		}
		for _, m := range p.Models {
			if m.ID == mid {
				return sel, nil
			}
		}
	}
	return "", fmt.Errorf("dsh: selected model %q is not projected by any provider", sel)
}

func splitSelection(sel string) (pid, mid string, err error) {
	i := strings.LastIndex(sel, "/")
	if i <= 0 || i == len(sel)-1 {
		return "", "", fmt.Errorf("dsh: selected model %q must be providerID/modelID", sel)
	}
	return sel[:i], sel[i+1:], nil
}

File: packages/clients/dsh/src/patch.ts
import { promises as fs } from 'node:fs';
import { join } from 'node:path';
import { DSH_BRIDGE_PLUGIN_FILENAME, DSH_BRIDGE_PLUGIN_SOURCE, DSH_BRIDGE_ROW_ID } from './bridge-asset.ts';
import { DSH_YOLO_PLUGIN_FILENAME, DSH_YOLO_PLUGIN_SOURCE, DSH_YOLO_ROW_ID } from './yolo-asset.ts';

/** File name of the generated loader overlay inside the per-run DSH home. */
export const DSH_LOADER_PATCH_FILENAME = 'wrenyard-loader-patch.yaml';

/** Exactly one Wrenyard Gateway provider is injected, keyed by its route. */
const DSH_GATEWAY_ROUTE_KEY = 'wrenyard';
const DSH_GATEWAY_PROVIDER_ROW_ID = 'llm-pi-ai';
const DSH_DEFAULT_MODEL_ROW_ID = 'agent-default-model';

/** The one credential name the generated overlay may reference; never a value. */
const DSH_GATEWAY_TOKEN_ENV = 'WRENYARD_GATEWAY_TOKEN';

/** Loader overlay insert id the native DSH loader uses for injected MCP clients. */
const DSH_MCP_CLIENT_ROW_ID = '@deepseek-ai/dsh-mcp-client';

/** The two MCP transports the DSH loader overlay can represent. */
export type DshMcpTransport = 'stdio' | 'streamable-http';

/**
 * A single projected MCP server for the DSH loader overlay. A stdio server
 * carries a command plus optional args/env; an HTTP server carries only its
 * transport. This mirrors the retired Go projection, which represented HTTP
 * servers by transport alone.
 */
export interface DshMcpServer {
  readonly name: string;
  readonly transport: DshMcpTransport;
  readonly command?: string;
  readonly url?: string;
  readonly headers?: Readonly<Record<string, string>>;
  readonly cwd?: string;
  readonly args?: readonly string[];
  /** Sorted `KEY=value` entries; a `!!js ` prefix is passed through unquoted. */
  readonly env?: readonly string[];
}

const YAML_PLAIN = /^[A-Za-z0-9][A-Za-z0-9._@/\-]*$/;

/** Render a YAML scalar, quoting only when the plain form is unsafe. */
function yamlStr(value: string): string {
  return YAML_PLAIN.test(value) ? value : JSON.stringify(value);
}

/**
 * Render a config value, passing an explicit `!!js ` tag through raw so a
 * credential reference stays unquoted rather than becoming a string literal.
 */
function yamlValue(value: string): string {
  return value.startsWith('!!js ') ? value : yamlStr(value);
}

/** Secret-free description of the single model route the overlay declares. */
export interface DshModelRoute {
  /** Selected model id, always exactly `provider/model`. */
  readonly publicModelId: string;
  /**
   * OpenAI-compatible chat base URL of the local Gateway. Present only for a
   * Wrenyard Gateway route; absent for a native route, which declares no
   * endpoint of ours because the native provider owns its own transport.
   */
  readonly baseUrl?: string;
  /** Optional context-window annotation for the declared model. */
  readonly contextWindow?: number;
}

/**
 * Render the complete secret-free loader overlay for one DSH invocation: the
 * `llm-pi-ai` provider row carrying the single Wrenyard Gateway route (gateway
 * routes only), the `agent-default-model` selection, the pinned execution mode,
 * and the two runtime insert rows that mount the transcript bridge and the
 * session normalizer.
 *
 * Only the selected model id is ever projected — the overlay never invents a
 * second model, and a native route declares no endpoint at all. The credential
 * is always the literal `!!js process.env.<ENV>` tag, so the generated file can
 * never hold a secret.
 *
 * `yoloPluginPath` and `bridgePluginPath` are absolute plugin paths, exactly as
 * the retired Go renderer required them; an empty bridge path omits its row.
 *
 * `mcpServers` are projected as `@deepseek-ai/dsh-mcp-client` insert rows after
 * the plugin rows, sorted by server name, matching the retired Go projection.
 */
export function renderDshPatch(route: DshModelRoute | undefined, yoloPluginPath: string, bridgePluginPath: string, mcpServers: readonly DshMcpServer[] = []): string {
  const lines: string[] = ['# wrenyard dsh patch (generated; secret-free)'];

  if (route) {
    const separator = route.publicModelId.indexOf('/');
    if (separator <= 0 || separator === route.publicModelId.length - 1)
      throw new Error(`dsh: model id must be provider/model: ${route.publicModelId}`);
    const providerId = route.publicModelId.slice(0, separator);
    const modelId = route.publicModelId.slice(separator + 1);
    if (route.baseUrl) {
      lines.push(
        `- id: ${yamlStr(DSH_GATEWAY_PROVIDER_ROW_ID)}`,
        '  config:',
        '    providers:',
        `      ${yamlStr(DSH_GATEWAY_ROUTE_KEY)}:`,
        '        displayName: Wrenyard',
        '        api: openai-completions',
        `        apiKeyEnv: ${yamlStr(DSH_GATEWAY_TOKEN_ENV)}`,
        `        baseURL: ${yamlStr(route.baseUrl)}`,
        '        models:',
        `          - id: ${yamlStr(route.publicModelId)}`,
        `            name: ${yamlStr(modelId)}`,
      );
      if (route.contextWindow && route.contextWindow > 0)
        lines.push(`            contextWindow: ${route.contextWindow}`);
    }
    lines.push(
      `- id: ${yamlStr(DSH_DEFAULT_MODEL_ROW_ID)}`,
      '  config:',
      `    provider: ${yamlStr(route.baseUrl ? DSH_GATEWAY_ROUTE_KEY : providerId)}`,
      `    model: ${yamlStr(route.baseUrl ? route.publicModelId : modelId)}`,
    );
  }

  // The generated overlay is always the final application layer. Keep the
  // enforcing services mounted, fix their defaults to YOLO, and remove the
  // obsolete user-facing permission selector.
  lines.push(
    '- id: sandbox-policy',
    '  config:',
    '    mode: danger-full-access',
    '- id: approval',
    '  config:',
    '    policy: never',
    '- id: permission',
    '  disabled: true',
    '- id: ui-permission',
    '  disabled: true',
  );
  // Mandatory session normalizer first, then the optional transcript bridge —
  // the exact insert order the retired Go patch renderer emitted.
  if (yoloPluginPath)
    lines.push(
      '- insert:',
      `    id: ${yamlStr(DSH_YOLO_ROW_ID)}`,
      `    name: ${yamlStr(yoloPluginPath)}`,
    );
  if (bridgePluginPath)
    lines.push(
      '- insert:',
      `    id: ${yamlStr(DSH_BRIDGE_ROW_ID)}`,
      `    name: ${yamlStr(bridgePluginPath)}`,
    );

  // MCP servers are inserted after the plugins, sorted by name. A transport the
  // loader cannot represent is rejected before any row is emitted.
  for (const server of [...mcpServers].sort((left, right) => left.name < right.name ? -1 : left.name > right.name ? 1 : 0)) {
    if (server.transport !== 'stdio' && server.transport !== 'streamable-http')
      throw new Error(`dsh: unsupported mcp transport ${JSON.stringify(server.transport)}`);
    lines.push(
      '- insert:',
      `    id: ${yamlStr('wrenyard-mcp-' + server.name)}`,
      `    name: ${yamlStr(DSH_MCP_CLIENT_ROW_ID)}`,
      '    config:',
      `      serverName: ${yamlStr(server.name)}`,
      `      transport: ${yamlStr(server.transport)}`,
      '      failOnStartupError: true',
    );
    if (server.transport === 'stdio') {
      lines.push(`      command: ${yamlStr(server.command ?? '')}`);
      if (server.cwd) lines.push(`      cwd: ${yamlStr(server.cwd)}`);
      if (server.args && server.args.length > 0) {
        lines.push('      args:');
        for (const arg of server.args)
          lines.push(`        - ${yamlStr(arg)}`);
      }
    }
    if (server.transport === 'streamable-http') {
      if (!server.url) throw new Error('DSH HTTP MCP server requires a URL');
      lines.push(`      url: ${yamlStr(server.url)}`);
      if (server.headers && Object.keys(server.headers).length > 0) {
        lines.push('      headers:');
        for (const [key, value] of Object.entries(server.headers))
          lines.push(`        ${yamlStr(key)}: ${yamlStr(value)}`);
      }
    }
    if (server.env && server.env.length > 0) {
      lines.push('      env:');
      for (const entry of [...server.env].sort()) {
        const separator = entry.indexOf('=');
        if (separator === -1)
          continue;
        lines.push(`        ${yamlStr(entry.slice(0, separator))}: ${yamlStr(entry.slice(separator + 1))}`);
      }
    }
  }

  return `${lines.join('\n')}\n`;
}

/**
 * Materialize the generated overlay and its two plugin assets into the isolated
 * per-run DSH home, returning the absolute overlay path. The caller owns the
 * home and removes it through its lifecycle cleanup.
 */
export async function writeLoaderPatch(home: string, route?: DshModelRoute, mcpServers: readonly DshMcpServer[] = []): Promise<string> {
  await fs.mkdir(home, { recursive: true });
  const bridgePath = join(home, DSH_BRIDGE_PLUGIN_FILENAME);
  const yoloPath = join(home, DSH_YOLO_PLUGIN_FILENAME);
  await fs.writeFile(bridgePath, DSH_BRIDGE_PLUGIN_SOURCE, { encoding: 'utf8', mode: 0o600 });
  await fs.writeFile(yoloPath, DSH_YOLO_PLUGIN_SOURCE, { encoding: 'utf8', mode: 0o600 });
  const target = join(home, DSH_LOADER_PATCH_FILENAME);
  await fs.writeFile(target, renderDshPatch(route, yoloPath, bridgePath, mcpServers), { encoding: 'utf8', mode: 0o600 });
  return target;
}
